brave/src/main/java/com/linecorp/armeria/common/brave/RequestContextCurrentTraceContext.java

```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.brave;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Collections;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import com.linecorp.armeria.client.brave.BraveClient;
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.server.brave.BraveService;
+
+import brave.Tracing;
+import brave.propagation.CurrentTraceContext;
+import brave.propagation.TraceContext;
+import io.netty.util.Attribute;
+import io.netty.util.AttributeKey;
+
+/**
+ * {@linkplain Tracing.Builder#currentTraceContext(CurrentTraceContext) Tracing
+ * context} implemented with {@link RequestContext}.
+ *
+ * <p>This {@link CurrentTraceContext} stores/loads the trace context into/from a
+ * {@link RequestContext}'s attribute so that there's no need for thread local variables
+ * which can lead to unpredictable behavior in asynchronous programming.
+ */
+public final class RequestContextCurrentTraceContext extends CurrentTraceContext {
+
+    /**
+     * Use this singleton when building a {@link Tracing} instance for use with
+     * {@link BraveService} or {@link BraveClient}.
+     *
+     * <p>If you need to customize the context, use {@link #builder()} instead.
+     *
+     * @see Tracing.Builder#currentTraceContext(CurrentTraceContext)
+     */
+    public static final CurrentTraceContext DEFAULT = new RequestContextCurrentTraceContext(new Builder());
```

Comment:
Should we go ahead and add a method `default()` instead of making the instance itself public?

Reply:
Sounds good. Let me do that. :-)